designs/helper_to_upgrade_projects_by_rescaffolding.md

```diff
@@ -0,0 +1,170 @@
+
+## Motivation
+
+A common scenario is to upgrade the project based on the newer Kubebuilder. The recommended (straightforward) steps are:
+
+- a) re-scaffold all files from scratch using the upper version/plugins
+- b) copy user-defined source code to the new layout
+
+The proposed command will automate the process at maximum, therefore helping operator authors with minimizing the manual effort.
+
+The main motivation of this proposal is to provide a helper for upgrades and
+make less painful this process. Examples:
+
+- See the discussion [How to regenerate scaffolding?](https://github.com/kubernetes-sigs/kubebuilder/discussions/2864)
+- From [slack channel By Paul Laffitte](https://kubernetes.slack.com/archives/CAR30FCJZ/p1675166014762669)
+
+### Goals
+
+- Help users upgrade their project with the latest changes
+- Help users to re-scaffold projects from scratch based on what was done previously with the tool
+- Make less painful the process to upgrade
+
+### Non-Goals
+
+- Change the default layout or how the KubeBuilder CLI works
+- Deal with customizations or deviations from the proposed layout
+- Be able to perform the project upgrade to the latest changes without human interactions
+- Deal and support external plugins
+- Provides support to [declarative](https://book.kubebuilder.io/plugins/declarative-v1.html) plugin
+  since it is desired and planned to decouple this solution and donate this plugin to its own authors [More info](https://github.com/kubernetes-sigs/kubebuilder/issues/3186)
+- Provide support old version prior we have the Project config (Kubebuilder < 3x) and the go/v2 layout which exist to ensure  a backwards compatibility with legacy layout provide by Kubebuilder 2x
+
+## Proposal
+
+The proposed solution to achieve this goal is to create an alpha command as described
+in the example section above, see:
+
+```shell
+kubebuilder alpha generate \
+    --from=<path of the project file> 
+    --to=<path where the project should be re-scaffold> 
+    --backup=<path-where the current version of the project should be copied as backup>
+    --plugins=<chain of plugins key that can be used to create the layout with init sub-command>
+```
+
+Where:
+
+- from: [Optional] If not informed then, the command would check it in the current directory
+- to: [Optional] If not informed then, it should be the current repository
+- backup: [Optional] If not informed then, the backup by copying the project to a new directory would not be made
+- plugins:  [Optional] If not informed then, it is the same plugin chain available in the layout field
+
+This command would mainly perform the following operations:
+
+- 1. Check the flags
+- 2. If the backup flag be used, then check if is a valid path and make a backup of the current project
+- 3. Ensure that the output path is clean
```

Comment:
```suggestion
- 3. Ensure that the output path is clean, if not error out.
```

This is to ensure that we end up producing a clean state of scaffolding.

Reply:
@varshaprasad96 I added on top in the example so workflows.
In this case, we cannot return error because by default this directory will not be empty.
It will have the files from the previous scaffold. So we need to remove all files. 

I also, improve this one to try make more clear.
Let me know wdyt?